src/http/http_serde.rs

```diff
@@ -0,0 +1,237 @@
+        .and_then(|header_value_str| header_value_str.parse().map_err(Into::into))
+}
+
+/// Header indicating the originating `HelperIdentity`.
+/// May be replaced in the future with a method with better security
+#[cfg_attr(feature = "enable-serde", derive(serde::Deserialize))]
+pub struct OriginHeader {
+    pub origin: HelperIdentity,
+}
+
+impl OriginHeader {
+    pub(crate) fn add_to(self, req: axum::http::request::Builder) -> axum::http::request::Builder {
+        req.header(ORIGIN_HEADER_NAME.clone(), self.origin)
+    }
+}
+
+#[async_trait]
+impl<B: Send> FromRequest<B> for OriginHeader {
+    type Rejection = server::Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let origin: usize = get_header(req, ORIGIN_HEADER_NAME.clone())?;
+        let origin = HelperIdentity::try_from(origin)
+            .map_err(|err| server::Error::InvalidHeader(err.into()))?;
+        Ok(OriginHeader { origin })
+    }
+}
+
+/// Headers that are expected on `Step` commands
+/// # `content_length`
+/// standard HTTP header representing length of entire body. Body length must be a multiple of
+/// `MESSAGE_PAYLOAD_SIZE_BYTES`
+/// # `offset`
+/// For any given batch, their `record_id`s must be known. The first record in the batch will have
+/// id `offset`, and subsequent records will be in-order from there.
+#[derive(Copy, Clone)]
+pub struct StepHeaders {
+    pub offset: u32,
+}
+
+impl StepHeaders {
+    pub(crate) fn add_to(self, req: axum::http::request::Builder) -> axum::http::request::Builder {
+        req.header(OFFSET_HEADER_NAME.clone(), self.offset)
+    }
+}
+
+#[async_trait]
+impl<B: Send> FromRequest<B> for StepHeaders {
+    type Rejection = server::Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let content_length: u32 = get_header(req, CONTENT_LENGTH_HEADER_NAME.clone())?;
+        let offset: u32 = get_header(req, OFFSET_HEADER_NAME.clone())?;
+        // content_length must be aligned with the size of an element
+        if content_length as usize % MESSAGE_PAYLOAD_SIZE_BYTES == 0 {
+            Ok(StepHeaders { offset })
+        } else {
+            Err(server::Error::WrongBodyLen {
+                body_len: content_length,
+                element_size: MESSAGE_PAYLOAD_SIZE_BYTES,
+            })
+        }
+    }
+}
+
+// there isn't an easy way to compose const strings at compile time, so we will hard-code
+// everything
+
+pub const ECHO_AXUM_PATH: &str = "/echo";
+
+pub fn echo_uri(payload: &str) -> String {
+    format!("/echo?foo={payload}")
+}
+
+// all queries
+pub const BASE_QUERY_AXUM_PATH: &str = "/query";
+
+// create query
+#[cfg_attr(feature = "enable-serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct CreateQueryResp {
+    pub query_id: QueryId,
+}
+
+pub const CREATE_QUERY_AXUM_PATH: &str = "/";
+
+pub fn create_query_uri(data: QueryConfig) -> String {
```

Comment:
maybe we should keep helper functions inside client module? this will likely grow and I'ld love to keep the shared state as small as possible. We certainly need to share URIs and very likely request/response (although I'd love to find a way not to do it).  

also maybe you can modularize it to add a bit more structure

```rust
mod query {
     mod create {
          struct Request { .. }, struct Response {.. }
    }
}
```

in this case you would know exactly what to look for for each handler

Reply:
that would mean the URI is in multiple places (one for server, one for client). if we are going to allow that, then would this basically end up looking the same as before?

Reply:
I attempted to implement the `struct Request` (`struct Response` wasn't really needed). take a look